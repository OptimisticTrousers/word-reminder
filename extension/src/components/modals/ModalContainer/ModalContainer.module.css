.modal {
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: var(--modal-bkg-color);
  z-index: 5;
}

.modal__container {
  border-radius: 8px;
  box-shadow: 4px 4px 20px rgb(0 0 0 / 30%);
  display: flex;
  flex-direction: column;
  min-height: min(100%, 428px);
  min-width: 310px;
  position: relative;
  width: min(100%, 500px);
}

.modal__header {
  margin-bottom: 0.5rem;
}

.modal__header button.modal__button {
  position: absolute;
  border-radius: 50%;
  background-color: transparent;
  right: 12px;
  top: 12px;
  padding: 0;
  color: var(--secondary-icon-color);
}

.modal__header button.modal__button:focus {
  outline: auto;
  box-shadow: none;
}

.modal__header .modal__button:hover {
  color: var(--tertiary-icon-color);
  transform: scale(1.075);
  background-color: transparent;
}

.modal__icon {
  height: 32px;
  width: 32px;
}

.modal__title-bar {
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 60px;
  border-bottom: 1px solid var(--break-color);
}

.modal__title {
  color: var(--text-color);
}
