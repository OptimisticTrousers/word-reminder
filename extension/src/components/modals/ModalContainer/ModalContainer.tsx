import { SquareX } from "lucide-react";
import { ReactNode } from "react";
import CSSModules from "react-css-modules";

import { useDisableScroll } from "../../../hooks/useDisableScroll";
import styles from "./ModalContainer.module.css";

interface Props {
  children: ReactNode;
  title: string;
  toggleModal: () => void;
}

export const ModalContainer = CSSModules(
  function ({ children, title, toggleModal }: Props) {
    useDisableScroll();

    function handleClose() {
      toggleModal();
    }

    return (
      <div
        styleName="modal"
        aria-modal="true"
        role="dialog"
        aria-labelledby="modal-title"
      >
        <div styleName="modal__container" className="scale-down">
          <header styleName="modal__header">
            <button
              styleName="modal__button"
              onClick={handleClose}
              aria-label="Close modal"
            >
              <SquareX styleName="modal__icon" />
            </button>
            <div styleName="modal__title-bar" id="modal-title">
              <h2 styleName="modal__title">{title}</h2>
            </div>
          </header>
          {children}
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: false, handleNotFoundStyleName: "log" }
);
