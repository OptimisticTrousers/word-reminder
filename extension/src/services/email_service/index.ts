export { emailService } from "./email_service";
