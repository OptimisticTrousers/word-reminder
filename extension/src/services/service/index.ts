export type { AuthParams, Params } from "./service";
export { service } from "./service";
