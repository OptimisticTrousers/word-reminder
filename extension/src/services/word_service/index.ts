export { wordService } from "./word_service";
