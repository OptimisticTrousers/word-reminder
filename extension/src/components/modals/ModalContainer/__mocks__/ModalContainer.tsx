import { Props } from "../ModalContainer";

export function ModalContainer({ title, toggleModal, children }: Props) {
  return (
    <div>
      <h2>{title}</h2>
      <button aria-label="Close modal" onClick={toggleModal}>
        X
      </button>
      <div>{children}</div>
    </div>
  );
}
