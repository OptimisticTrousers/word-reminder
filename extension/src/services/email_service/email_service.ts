import { Templates } from "common";

import { service } from "../service";

interface SendEmailBody {
  email: string;
  subject: string;
  template: Templates;
}

interface VerifyEmailCodeBody {
  code: string;
}

export const emailService = (function (service) {
  const { post, VITE_API_DOMAIN } = service;

  function sendEmail(body: SendEmailBody) {
    return post({
      url: `${VITE_API_DOMAIN}/emails`,
      options: {
        body: JSON.stringify(body),
        credentials: "include",
      },
    });
  }

  function verifyEmailCode(body: VerifyEmailCodeBody) {
    return post({
      url: `${VITE_API_DOMAIN}/emails`,
      options: {
        body: JSON.stringify(body),
        credentials: "include",
      },
    });
  }

  return { sendEmail, verifyEmailCode };
})(service);
