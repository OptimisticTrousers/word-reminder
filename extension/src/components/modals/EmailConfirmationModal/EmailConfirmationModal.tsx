import { CODE_MAX_BYTES, User } from "common";
import { useContext } from "react";
import CSSModules from "react-css-modules";
import { useOutletContext } from "react-router-dom";
import { useMutation } from "@tanstack/react-query";

import {
  NotificationContext,
  NOTIFICATION_ACTIONS,
} from "../../../context/Notification";
import { useNotificationError } from "../../../hooks/useNotificationError";
import { ModalContainer } from "../ModalContainer";
import { emailService } from "../../../services/email_service";
import { ErrorResponse } from "../../../types";
import styles from "./EmailConfirmationModal.module.css";

interface Props {
  toggleModal: () => void;
}

export const EmailConfirmationModal = CSSModules(
  function ({ toggleModal }: Props) {
    const { user }: { user: User } = useOutletContext();
    const { showNotification } = useContext(NotificationContext);
    const { showNotificationError } = useNotificationError();
    const { isPending, mutate } = useMutation({
      mutationFn: emailService.verifyEmailCode,
      onError: (response: ErrorResponse) => {
        showNotificationError(response);
      },
      onSuccess: () => {
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          EMAIL_CONFIRMATION_NOTIFICATION_MSGS.confirmEmail(user.email)
        );
      },
      onSettled: () => {
        toggleModal();
      },
    });

    function handleConfirm(formData: FormData) {
      const body = {
        code: formData.get("code") as string,
      };
      mutate(body);
    }

    function handleCancel() {
      toggleModal();
    }

    return (
      <ModalContainer title="Confirm your email" toggleModal={toggleModal}>
        <form styleName="modal" action={handleConfirm}>
          <p styleName="modal__message">
            Please enter the confirmation code that was sent to{" "}
            <span styleName="modal__bold">{user.email}</span> within 5 minutes.
          </p>
          <label styleName="modal__label">
            Code
            <input
              styleName="modal__input"
              name="code"
              type="text"
              required={true}
              disabled={isPending}
              maxLength={CODE_MAX_BYTES}
            />
          </label>
          <div styleName="modal__buttons">
            <button
              styleName="modal__button modal__button--cancel"
              onClick={handleCancel}
              disabled={isPending}
            >
              Cancel
            </button>
            <button
              styleName="modal__button modal__button--confirm"
              type="submit"
              disabled={isPending}
            >
              Enter Code
            </button>
          </div>
        </form>
      </ModalContainer>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "ignore" }
);

const EMAIL_CONFIRMATION_NOTIFICATION_MSGS = {
  confirmEmail: (email: string) => {
    return `You have confirmed your email: ${email}`;
  },
};
