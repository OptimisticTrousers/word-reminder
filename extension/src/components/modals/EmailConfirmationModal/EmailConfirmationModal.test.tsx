import { createRoutesStub, Outlet } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { Props } from "../ModalContainer/ModalContainer";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import userEvent from "@testing-library/user-event";
import { emailService } from "../../../services/email_service";
import { EmailConfirmationModal } from "./EmailConfirmationModal";

vi.mock("../ModalContainer", () => {
  return {
    ModalContainer: function ({ title, toggleModal, children }: Props) {
      return (
        <div>
          <h2>{title}</h2>
          <button aria-label="Close modal" onClick={toggleModal}>
            X
          </button>
          <div>{children}</div>
        </div>
      );
    },
  };
});
describe("EmailConfirmationModal", () => {
  const testUser = {
    email: "bob@email.com",
    id: "1",
  };

  const info = {
    message:
      "Message sent: <d786aa62-4e0a-070a-47ed-0b0666549519@ethereal.email",
  };
  const status = 200;

  const mockToggleModal = vi.fn();

  function setup() {
    const queryClient = new QueryClient();
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/wordReminders",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <EmailConfirmationModal toggleModal={mockToggleModal} />
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/wordReminders"]} />),
    };
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  // describe("send confirmation modal", () => {
  //   it("shows form", async () => {
  //     const { asFragment } = setup();

  //     const message = screen.getByText(
  //       "In order to verify it is you, we need to send an email associated with this account."
  //     );
  //     expect(message).toBeInTheDocument();
  //     expect(asFragment()).toMatchSnapshot();
  //   });

  //   it("calls the functions to send a confirmation email", async () => {
  //     const mockSendEmail = vi
  //       .spyOn(emailService, "sendEmail")
  //       .mockImplementation(async () => {
  //         return { json: info, status };
  //       });

  //     const { user } = setup();

  //     const sendConfirmationButton = screen.getByRole("button", {
  //       name: "Send Confirmation",
  //     });
  //     await user.click(sendConfirmationButton);

  //     const modal = screen.queryByRole("dialog");
  //     expect(modal).not.toBeInTheDocument();
  //     expect(mockSendEmail).toHaveBeenCalledTimes(1);
  //     expect(mockSendEmail).toHaveBeenCalledWith({
  //       email: testUser.email,
  //       subject: "Confirm Your Email",
  //       template: Templates.EMAIL_VERIFICATION,
  //     });
  //     expect(mockToggleModal).not.toHaveBeenCalled();
  //   });

  //   it("calls the functions to show notification error", async () => {
  //     const message = "Error Message.";
  //     const status = 400;
  //     const mockSendEmail = vi
  //       .spyOn(emailService, "sendEmail")
  //       .mockImplementation(async () => {
  //         return Promise.reject({ json: { message }, status });
  //       });

  //     const { user } = setup();
  //     const sendConfirmationButton = screen.getByRole("button", {
  //       name: "Send Confirmation",
  //     });
  //     await user.click(sendConfirmationButton);

  //     const modal = screen.queryByRole("dialog");
  //     expect(mockSendEmail).toHaveBeenCalledTimes(1);
  //     expect(mockSendEmail).toHaveBeenCalledWith({
  //       email: testUser.email,
  //       subject: "Confirm Your Email",
  //       template: Templates.EMAIL_VERIFICATION,
  //     });
  //     expect(modal).not.toBeInTheDocument();
  //     expect(mockToggleModal).not.toHaveBeenCalled();
  //   });

  //   it("changes the modal so the user can enter the code", async () => {
  //     vi.spyOn(emailService, "sendEmail").mockImplementation(async () => {
  //       return { json: info, status };
  //     });
  //     const { user } = setup();

  //     const sendConfirmationButton = screen.getByRole("button", {
  //       name: "Send Confirmation",
  //     });
  //     await user.click(sendConfirmationButton);

  //     const enterCodeButton = screen.getByRole("button", {
  //       name: "Enter Code",
  //     });
  //     const message = screen.getByText(
  //       `Please enter the code that was sent to your email: ${testUser.email}.`
  //     );
  //     expect(enterCodeButton).toBeInTheDocument();
  //     expect(message).toBeInTheDocument();
  //     expect(mockToggleModal).not.toHaveBeenCalled();
  //   });

  //   it("user closes the modal after modal was changed", async () => {
  //     vi.spyOn(emailService, "sendEmail").mockImplementation(async () => {
  //       return { json: info, status };
  //     });
  //     const { user } = setup();
  //     const sendConfirmationButton = screen.getByRole("button", {
  //       name: "Send Confirmation",
  //     });
  //     await user.click(sendConfirmationButton);

  //     const cancelButton = screen.getByRole("button", { name: "Cancel" });
  //     await user.click(cancelButton);

  //     const modal = screen.queryByRole("dialog");
  //     expect(modal).not.toBeInTheDocument();
  //     expect(mockToggleModal).toHaveBeenCalledTimes(1);
  //     expect(mockToggleModal).toHaveBeenCalledWith();
  //   });

  //   it("user closes the modal using cancel button", async () => {
  //     const { user } = setup();

  //     const cancelButton = screen.getByRole("button", { name: "Cancel" });
  //     await user.click(cancelButton);

  //     const modal = screen.getByRole("dialog");
  //     expect(modal).not.toBeInTheDocument();
  //     expect(mockToggleModal).toHaveBeenCalledTimes(1);
  //     expect(mockToggleModal).toHaveBeenCalledWith();
  //   });

  //   it("user closes the modal using x button", async () => {
  //     const { user } = setup();

  //     const modalContainerCloseButton = screen.getByRole("button", {
  //       name: "Close modal",
  //     });
  //     await user.click(modalContainerCloseButton);

  //     const modal = screen.getByRole("dialog");
  //     expect(modal).not.toBeInTheDocument();
  //     expect(mockToggleModal).toHaveBeenCalledTimes(1);
  //     expect(mockToggleModal).toHaveBeenCalledWith();
  //   });
  // });

  describe("type email code modal", () => {
    it("shows form", async () => {
      const { asFragment } = setup();

      const message = screen.getByText((_, element) => {
        return (
          element?.textContent ===
          `Please enter the confirmation code that was sent to ${testUser.email} within 5 minutes.`
        );
      });
      expect(message).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("calls the functions to verify confirmation email code", async () => {
      const mockVerifyEmailCode = vi
        .spyOn(emailService, "verifyEmailCode")
        .mockImplementation(async () => {
          return { json: info, status };
        });
      const { user } = setup();

      const code = "code";
      const codeInput = screen.getByLabelText("Code", { selector: "input" });
      await user.type(codeInput, code);
      const enterCodeButton = screen.getByRole("button", {
        name: "Enter Code",
      });
      await user.click(enterCodeButton);

      expect(mockVerifyEmailCode).toHaveBeenCalledTimes(1);
      expect(mockVerifyEmailCode).toHaveBeenCalledWith({ code });
      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });

    it("does not allow the user to submit when the code field is empty", async () => {
      const mockVerifyEmailCode = vi
        .spyOn(emailService, "verifyEmailCode")
        .mockImplementation(async () => {
          return { json: info, status };
        });

      const { user } = setup();

      const enterCodeButton = screen.getByRole("button", {
        name: "Enter Code",
      });
      await user.click(enterCodeButton);

      expect(mockVerifyEmailCode).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
    });

    it("calls the functions to show notification error", async () => {
      const message = "Error Message.";
      const status = 400;
      const mockVerifyEmailCode = vi
        .spyOn(emailService, "verifyEmailCode")
        .mockImplementation(async () => {
          return Promise.reject({ json: { message }, status });
        });
      const { user } = setup();

      const code = "Wrong code.";
      const codeInput = screen.getByLabelText("Code", { selector: "input" });
      await user.type(codeInput, code);
      const enterCodeButton = screen.getByRole("button", {
        name: "Enter Code",
      });
      await user.click(enterCodeButton);

      expect(mockVerifyEmailCode).toHaveBeenCalledTimes(1);
      expect(mockVerifyEmailCode).toHaveBeenCalledWith({ code });
      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });

    it("user closes the modal using cancel button", async () => {
      const { user } = setup();

      const cancelButton = screen.getByRole("button", { name: "Cancel" });
      await user.click(cancelButton);

      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });

    it("user closes the modal using x button", async () => {
      const { user } = setup();

      const modalContainerCloseButton = screen.getByRole("button", {
        name: "Cancel",
      });
      await user.click(modalContainerCloseButton);

      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });
  });
});
