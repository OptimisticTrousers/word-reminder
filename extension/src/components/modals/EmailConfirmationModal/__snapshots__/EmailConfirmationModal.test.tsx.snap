// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailConfirmationModal > type email code modal > shows form 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Confirm your email
    </h2>
    <button
      aria-label="Close modal"
    >
      X
    </button>
    <div>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal_516802"
      >
        <p
          class="_modal__message_516802"
        >
          Please enter the confirmation code that was sent to 
          <span
            class="_modal__bold_516802"
          >
            bob@email.com
          </span>
           within 5 minutes.
        </p>
        <label
          class="_modal__label_516802"
        >
          Code
          <input
            class="_modal__input_516802"
            maxlength="16"
            name="code"
            required=""
            type="text"
          />
        </label>
        <div
          class="_modal__buttons_516802"
        >
          <button
            class="_modal__button_516802 _modal__button--cancel_516802"
          >
            Cancel
          </button>
          <button
            class="_modal__button_516802 _modal__button--confirm_516802"
            type="submit"
          >
            Enter Code
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
